### PR TITLE
First iteration of function hooks

### PR DESCRIPTION
**Pull request by kvalakuzhyzp**
One thing to fix for this.
1) Store the function information in an OS agnostic way.

### CODE REVIEW

File: src/zelos/hooks.py [hunk 1]
```diff
@@ -287,6 +294,69 @@ def syscall_callback_wrapper(zelos, sysname, args, retval):
...
+            self._func_hooks_to_register.append((name, callback))
+            return
+
+        address_ptr += self.z.memory.get_module_base(self.z.target_binary_path)
+
+        def read_wrapper(z, access, address, size, value):
+            callback(z)
+
+        # TODO: Another way to do this is to wait for these pointers
+        # to be written to, and then hook those execution locations.
```
Comment (rcourt) [inline src/zelos/hooks.py:348]: Would this work by creating a __write hook__ on `address_ptr`, the callback of which would create the __exec hook__   (just like the inline comment below)?

Comment (rcourt) [reply src/zelos/hooks.py:348]: Resolved offline.

File: tests/test_hook_manager.py [hunk 2]
```diff
@@ -291,6 +291,23 @@ def inst_hook(zelos, address, size):
         z.start()
         self.assertEqual(z.regs.getIP(), 0x8048B72)
 
+    def test_func_hook(self):
```
Comment (rcourt) [inline tests/test_hook_manager.py:294]: Include another test that hooks a function in the middle of a run, after the target module has been loaded into memory. For example, hook `malloc` after it has already been called once and see if the remaining call to malloc is hooked successfully.

Comment (kvalakuzhyzp) [reply tests/test_hook_manager.py:294]: Addressed in an update to the test. Added ability to set a condition to delete the hook as well.

**Pull request merged**